### ISSUE TITLE
Support decompressing IMPLODE-d files

### ISSUE THREAD

**Issue by bengarrett** [jvilk/BrowserFS#141]
I've been using The Emularity and EM-DOSBox on a website for DOSBox emulation in the browser. The Emularity has BrowserFS as a dependency so it can mount zip archives into EM-DOSBox and trick it into thinking the archive is a local file directory.

I've noticed that some zip archives do not work correctly. While they will mount and list the package's content some files within will not be accessible for read and write operations. I believe it might be an issue with BrowserFS?

I ran PKZip's `unzip -Zv` in DOSBox on a few of the offending files and they all share the same issue. Files compressed with the _imploded 1.0_ method are the ones that seem to fail.

For example in cncmag02.zip, ASCMAG.EXE works but _MENU_PL.EXE fails.
[You can test in the DOS emulator here](https://defacto2.net/file/detail/a56d9?dosuseoriginalzip=true&dosautorun=false) and also download the zip archive.

![cncmag02.zip](http://i.imgur.com/fwoJJuL.png)

Partial `unzip -Zv cncmag02.zip` output

```
Central directory entry #5:

  _MENU_PL.EXE

  offset of local header from start of archive:   194500 (0002F7C4h) bytes
  file system or operating system of origin:      MS-DOS, OS/2 or NT FAT
  version of encoding software:                   2.0
  minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
  minimum software version required to extract:   1.0
  compression method:                             imploded
  size of sliding dictionary (implosion):         4K
  number of Shannon-Fano trees (implosion):       2
  file security status:                           not encrypted
  extended local header:                          no
  file last modified on (DOS date/time):          1994 May 24 00:00:16
  32-bit CRC value (hex):                         af2fe293
  compressed size:                                2138 bytes
  uncompressed size:                              3546 bytes
  length of filename:                             12 characters
  length of extra field:                          0 bytes
  length of file comment:                         0 characters
  disk number on which file begins:               disk 1
  apparent file type:                             binary
  non-MSDOS external file attributes:             000000 hex
  MS-DOS file attributes (20 hex):                arc 

  There is no file comment.

Central directory entry #6:

  ASCMAG.EXE

  offset of local header from start of archive:   196680 (00030048h) bytes
  file system or operating system of origin:      MS-DOS, OS/2 or NT FAT
  version of encoding software:                   2.0
  minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
  minimum software version required to extract:   1.0
  compression method:                             none (stored)
  file security status:                           not encrypted
  extended local header:                          no
  file last modified on (DOS date/time):          1996 Jan 31 23:15:38
  32-bit CRC value (hex):                         cc712d24
  compressed size:                                15284 bytes
  uncompressed size:                              15284 bytes
  length of filename:                             10 characters
  length of extra field:                          0 bytes
  length of file comment:                         0 characters
  disk number on which file begins:               disk 1
  apparent file type:                             binary
  non-MSDOS external file attributes:             000000 hex
  MS-DOS file attributes (20 hex):                arc 

  There is no file comment.
```

[tsan-1.zip](https://defacto2.net/file/detail/a6549?dosuseoriginalzip=true&dosautorun=false)

![tsan-1.zip](http://i.imgur.com/NHCa21U.png?)

`unzip -Zv tsan-1.zip` output

```
Archive:  TSAN-1.ZIP
The zipfile comment is 492 bytes long and contains the following text:
======================== zipfile comment begins ==========================

Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°
Û²±°       This File Courtesy of ... R i v e n d e l l          Û²±°
Û²±°             2 0 1  -  8 0 8  -  5 5 7 4                    Û²±°
Û²±°  Running Revelation v0.50 Alpha - Author Site (Force Ten)  Û²±°
Û²±°         66oMB On-Line / No NUP / 144oo v.42bis HST         Û²±°
Û²±° TSAN/RR-INC/MaD EHQ/CWF/PVS/LoDD/MAi - TCSNet - iNSANiTY   Û²±°
Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°Û²±°
========================= zipfile comment ends ===========================

End-of-central-directory record:
-------------------------------

  Zip archive file size:                    160321 (00027241h)
  Actual end-cent-dir record offset:        159807 (0002703Fh)
  Expected end-cent-dir record offset:      159807 (0002703Fh)
  (based on the length of the central directory and its expected offset)

  This zipfile constitutes the sole disk of a single-part archive; its
  central directory contains 3 entries.
  The central directory is 192 (000000C0h) bytes long,
  and its (expected) offset in bytes from the beginning of the zipfile
  is 159615 (00026F7Fh).


Central directory entry #1:
---------------------------

  TSAN.EXE

  offset of local header from start of archive:   0 (00000000h) bytes
  file system or operating system of origin:      MS-DOS, OS/2 or NT FAT
  version of encoding software:                   1.1
  minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
  minimum software version required to extract:   1.0
  compression method:                             imploded
  size of sliding dictionary (implosion):         4K
  number of Shannon-Fano trees (implosion):       2
  file security status:                           not encrypted
  extended local header:                          no
  file last modified on (DOS date/time):          1991 Nov 17 03:55:08
  32-bit CRC value (hex):                         e441c5d1
  compressed size:                                10910 bytes
  uncompressed size:                              24576 bytes
  length of filename:                             8 characters
  length of extra field:                          26 bytes
  length of file comment:                         0 characters
  disk number on which file begins:               disk 1
  apparent file type:                             binary
  non-MSDOS external file attributes:             000000 hex
  MS-DOS file attributes (20 hex):                arc 

  The central-directory extra field contains:
  - A subfield with ID 0x0007 (PKWARE AV) and 22 data bytes:
    63 38 f6 2b 58 1f 86 36 ff d2 68 6a 03 c4 c8 f9 e1 7a 5e 8f f7 73.

  There is no file comment.

Central directory entry #2:
---------------------------

  TSAN.DAT

  offset of local header from start of archive:   10974 (00002ADEh) bytes
  file system or operating system of origin:      MS-DOS, OS/2 or NT FAT
  version of encoding software:                   1.1
  minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
  minimum software version required to extract:   1.0
  compression method:                             imploded
  size of sliding dictionary (implosion):         4K
  number of Shannon-Fano trees (implosion):       2
  file security status:                           not encrypted
  extended local header:                          no
  file last modified on (DOS date/time):          1991 Nov 10 01:21:18
  32-bit CRC value (hex):                         35ac2200
  compressed size:                                146397 bytes
  uncompressed size:                              293931 bytes
  length of filename:                             8 characters
  length of extra field:                          0 bytes
  length of file comment:                         0 characters
  disk number on which file begins:               disk 1
  apparent file type:                             binary
  non-MSDOS external file attributes:             000000 hex
  MS-DOS file attributes (20 hex):                arc 

  There is no file comment.

Central directory entry #3:
---------------------------

  RVNDL-ML.ICE

  offset of local header from start of archive:   157409 (000266E1h) bytes
  file system or operating system of origin:      MS-DOS, OS/2 or NT FAT
  version of encoding software:                   1.1
  minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
  minimum software version required to extract:   1.0
  compression method:                             imploded
  size of sliding dictionary (implosion):         4K
  number of Shannon-Fano trees (implosion):       2
  file security status:                           not encrypted
  extended local header:                          no
  file last modified on (DOS date/time):          1991 Sep 14 20:40:12
  32-bit CRC value (hex):                         304a919a
  compressed size:                                2164 bytes
  uncompressed size:                              6903 bytes
  length of filename:                             12 characters
  length of extra field:                          0 bytes
  length of file comment:                         0 characters
  disk number on which file begins:               disk 1
  apparent file type:                             binary
  non-MSDOS external file attributes:             000000 hex
  MS-DOS file attributes (20 hex):                arc 

  There is no file comment.

```

**Comment by jvilk** [comment]
Aha! It's as I feared: There are zip archives lurking in the past that use non-DEFLATE-based compression.

If there's a JavaScript implementation of implode (or if we can emscriptify one), we can slot it into [this switch statement](https://github.com/jvilk/BrowserFS/blob/master/src/backend/ZipFS.ts#L238).

**Comment by jvilk** [comment]
Here's an implementation of EXPLODE, which we could either try to Emscriptify, or independently re-implement:

ftp://ftp.cs.ait.ac.th/pub/archiver/unzip/explode.c

[Gist version](https://gist.github.com/jvilk/d1494ffb3d04df350c7d)

**Comment by jvilk** [comment]
This isn't something I'd have time to work on, but I most definitely accept PRs if you or someone else wants to take a stab at it!

**Comment by bengarrett** [comment]
I understand, it seems like a bit of work.

It appears that the deflate method wasn't introduced until 1993 with PKZIP 2. So there were at least 4 years during the peak of the BBS years, where PKZIP 1.x's implode, shrinking and crunching methods were in common use. It would probably be a longer period as I'd imagine people distributing software would use PKZIP 1 over 2 for better end user compatibility. Much like in cncmag02.zip example above where they compressed the archive using PKZIP 2 but used version 1's methods.

It would be really nice at some stage to have legacy method 2 to 6 extraction translated into JavaScript. But porting decompression methods into JS is well beyond my ability. But I will play around with Emscripten when I get a chance. 

Right now my workaround is to flag and automatically convert any legacy archives into deflate compatible zips on the back-end before they are loaded by Browser.JS.

But if anyone out there is keen here is a starting point.
https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT

> 4.4.5 compression method: (2 bytes)
        ...
        2 - The file is Reduced with compression factor 1
        3 - The file is Reduced with compression factor 2
        4 - The file is Reduced with compression factor 3
        5 - The file is Reduced with compression factor 4
        6 - The file is Imploded
 
...
> 5.1 UnShrinking - Method 1
    5.1.1 Shrinking is a Dynamic Ziv-Lempel-Welch compression algorithm
    with partial clearing.  The initial code size is 9 bits, and the 
    maximum code size is 13 bits.  Shrinking differs from conventional 
    Dynamic Ziv-Lempel-Welch implementations in several respects:

...
> 5.2 Expanding - Methods 2-5
    5.2.1 The Reducing algorithm is actually a combination of two
    distinct algorithms.  The first algorithm compresses repeated
    byte sequences, and the second algorithm takes the compressed
    stream from the first algorithm and applies a probabilistic
    compression method.

...
> 5.3 Imploding - Method 6
    5.3.1 The Imploding algorithm is actually a combination of two 
    distinct algorithms.  The first algorithm compresses repeated byte
    sequences using a sliding dictionary.  The second algorithm is
    used to compress the encoding of the sliding dictionary output,
    using multiple Shannon-Fano trees.

**Comment by jvilk** [comment]
For funsies, I tried my hand at a direct port of Info-Zip's EXPLODE the other day (method 6), but it doesn't quite work yet. My time limit on debugging it expired, and I returned to more pressing work.

[You can see it here.](https://github.com/jvilk/decompress-implode)

Maybe you or someone else can find my bug!

**Comment by jvilk** [comment]
@bengarrett After some quick debugging while in transit today, I got my EXPLODE implementation to work! But... I need some test files that exercise all of the main code paths so I know it truly works well.

The options that need to be tested:

* 4K dictionary, no literal tree: `cncmag02.zip`
* 8K dictionary, no literal tree: Need sample file
* 4K dictionary, literal tree: Need sample file
* 8K dictionary, literal tree: Need sample file

Do you know if you have any files in your collection with different #s of trees / dictionary sizes for IMPLODE?

**Comment by jvilk** [comment]
Alternatively, and probably preferably, maybe we could figure out how to create new ZIP files with these settings?

**Comment by bengarrett** [comment]
It seems PKZIP doesn't give much control over the methods of compression. I managed to source PKZIP 0.8 through to 2.04 and they each have different implementations.

0.8-0.92 allow you to force 'reduce' compression levels 1-4 on both ascii and binary which gives you 8 options.

1.x dropped support for reduce and only allows you to state whether you want to use 'shrink' or 'implode' methods using the `-ex` and `-es` options but gives you no control at all.

2.04 only implemented deflate for compression but gave you a few levels of control.
```
-ex	    Maximum compression (most compression, most time)
-en	    Normal compression (default)
-ef	    Fast compression
-es	    Super Fast compression (least compression, least time)
-e0	    Zero compression (storing)
```

So most of these are probably not going to be useful but I created a collection of ZIP archives using all the different compression methods and options available with PKZIP 0.8 - 2.04.

[pkzip_example_methods.zip](https://github.com/jvilk/BrowserFS/files/181013/pkzip_example_methods.zip)

**Comment by jvilk** [comment]
Awesome; thank you so much! Did you use data that I can legally include in the repository? I want to use these as test fixtures.

As a side note, it looks like Info-Zip also supports `unshrink`, and the source code looks significantly easier to port than `explode`. I haven't located `expand`/`unreduce` yet.

**Comment by bengarrett** [comment]
No unfortunately they would not be legal to include. I was lazy and just zipped the pkzip program directories themselves. As they were convenient because they contained a good mix of binaries and ascii text files.

**Comment by jvilk** [comment]
Aw darn. Any chance I could get you to compress something in the public domain? Maybe a Project Gutenberg book along with a randomly generated binary blob?

**Comment by bengarrett** [comment]
If you need data that you could include in the repo I could probably compress parts of FreeDOS which are GPL2. http://www.freedos.org/software/?cat=util

**Comment by jvilk** [comment]
That sounds fine. The license of the test fixtures won't affect the license of the actual library code AFAIK, so GPLv2 is OK!

**Comment by jvilk** [comment]
I took a few minutes port over unshrink and unreduce, and they work! They're in the EXPLODE repository, and the extract script can handle them.

When I have time, I'll publish all of them as separate NPM modules.

One caveat: unreduce apparently has a restrictive license; you cannot use it for commercial use, and must publish the source code of anything using it. Info-Zip actually doesn't distribute it by default. This is likely why you cannot find unreduce in most ZIP implementations.

**Comment by bengarrett** [comment]
That's brilliant news. 

Would maybe this be a suitable GPL2 alternative for unreduce?
http://aluigi.altervista.org/papers/unreduce.c

**Comment by jvilk** [comment]
@bengarrett That works. I just switched over to that version.

**Comment by bengarrett** [comment]
Great that fixes that problem! I've attached a newly compiled collection of PKZip 0.x, 1.x & 2.x archives for DOS. These ZIPs contain a mixture of binaries and ASCII example files that I created for testing in a previous project. You are welcome to include them in the BrowserFS repo and release them under whichever compatible licence you choose.

[pkzip-for-dos-examples-of-methods.zip](https://github.com/jvilk/BrowserFS/files/183265/pkzip-for-dos-examples-of-methods.zip)
